### ISSUE TITLE
should installing workers be persisted if browser is killed?

### ISSUE THREAD

**Issue by wanderview** [slightlyoff/ServiceWorker#794]
Consider the situation where a new SW is being installed.  Its been set as registration.installing and is in the process of handling the install event.  Before install completes, however, the browser is killed.

In this case, what should happen on browser restart?  Is registration.installing still set?  If so, when do we re-run the install event?

Its not clear to me the update mechanism will account for this case since the script will be a byte-for-byte match.

Personally I would be ok simply not persisting .installing workers.  Only workers that make it to .waiting would be persistent across browser restarts.

Thoughts?

**Comment by wanderview** [comment]
@matto What does chrome do in this case?

**Comment by jungkees** [comment]
> In this case, what should happen on browser restart? Is registration.installing still set? If so, when do we re-run the install event?

registration.installing was set and persisted but will be set to null in the next Update (step 4.2.3).

> Its not clear to me the update mechanism will account for this case since the script will be a byte-for-byte match.

When the execution reaches the byte-for-byte check, registration.installing should have been null (by the above) and will go on to the step 4.26.

**Comment by mattto** [comment]
Chrome simply doesn't persist .installing workers, as @wanderview proposed.

**Comment by jungkees** [comment]
> Personally I would be ok simply not persisting .installing workers. Only workers that make it to .waiting would be persistent across browser restarts.

I'm not sure how to spec that behavior if desirable.

The behavior of the spec and the impl is mostly aligned except a subtle difference that the spec requires `getRegistration()` on the restart should return non-null `registration.installing` whereas the impl returns null.

**Comment by mattto** [comment]
Well, Chrome is even more different: `getRegistration()` itself returns null if there was never a successfully installed worker for that registration. In Chrome's design all persisted registrations have exactly one persisted worker (a waiting one or an active one).

I think that's cleaner, that way you don't need to keep around state for half-installed workers. It also applies to cases like a disk error while trying to write the worker.

**Comment by jungkees** [comment]
I think there should be certain cases where clients want to get access to `.installing`, `.waiting` and `.active` at the same time, e.g. messaging between them. The whole point we divided the states of the workers that can be running in parallel is to maintain the versions of the workers without breaking the state.

**Comment by mattto** [comment]
Right, Chrome supports that but only one worker is persisted on disk. The .installing worker is written to temp storage and thrown away if it never reached .waiting.

**Comment by jungkees** [comment]
Understood. Also, agreed on the point that the .installing never reached .waiting before browser termination is sort of a garbage that doesn't get a chance to be installed again. I think, for the spec, garbage collecting those registrations having only the .installing when browser is killed can be an idea.

**Comment by jungkees** [comment]
One thing to check is the scenario where a registration has both .active and .waiting, and the browser is killed in that state. When the browser restarts and the .active spins off on an event and invokes `clients.claim()`. I think this should be working. For this scenario, shouldn't both .active and .waiting be persisted?

**Comment by mattto** [comment]
Chrome's position is that when the browser is closed, the .active worker no longer controls a client. So it restarts with only the .waiting worker which is activated on the next fetch or push event.

**Comment by wanderview** [comment]
I like the idea of persisting only one service worker on disk.  It simplifies things considerably.

From a spec language point of view, could we have a section like "Browser Shutdown" that says which of installing/waiting/active get discarded and which get persisted to next browser start?

**Comment by wanderview** [comment]
I wrote a gecko bug for us to simplify our persisted information to match chrome's behavior: https://bugzilla.mozilla.org/show_bug.cgi?id=1229795

**Comment by jungkees** [comment]
> could we have a section like "Browser Shutdown" that says which of installing/waiting/active get discarded and which get persisted to next browser start?

Agreed.

> Chrome's position is that when the browser is closed, the .active worker no longer controls a client. So it restarts with only the .waiting worker which is activated on the next fetch or push event.

I think the next fetch when the browser restarts will not activate the .waiting. The fetch just falls back to the network. (There's no active worker to dispatch fetch event at, so it doesn't get a chance to invoke Soft Update either.) So, making "Browser Shutdown" invoke Activate for all the registrations having both .active and .waiting makes sense? Let me know if this conforms to the current Chrome behavior.

**Comment by mattto** [comment]
> So, making "Browser Shutdown" invoke Activate for all the registrations having both .active and .waiting makes sense? Let me know if this conforms to the current Chrome behavior.

One difference is when you restart Chrome, if you getRegistration(scope) from an out of scope page you can get a registration with non-null waiting and null active.

That could probably be considered an implementation bug though. On getRegistration Chrome could trigger Activate that case, then I think it'd conform to the spec.

**Comment by wanderview** [comment]
So we need to say anything about dispatching the activate event on browser start when promoting a waiting worker?  Maybe we should just queue an Activate job?
